simulator.go

```diff
@@ -202,13 +204,46 @@ type simulatorAPIHandler struct {
 	overrides []string
 	autoID    uint32
 
-	runner *docker.Container
-	nodes  map[string]*docker.Container
+	runner       *docker.Container
+	nodes        map[string]*docker.Container
+	nodesTimeout map[string]time.Time
 
 	result *simulationResult
 	lock   sync.RWMutex
 }
 
+// CheckTimeout is a goroutine that checks if the timeout has passed and stops
+// container if it has.
+func (h *simulatorAPIHandler) CheckTimeout() {
+	for {
+		for id, c := range h.nodes {
+			h.lock.Lock()
+			if !c.State.Running || (time.Now().After(h.nodesTimeout[id])) {
+				h.terminateContainer(id, nil)
+			}
+			h.lock.Unlock()
+		}
+		time.Sleep(timeoutCheckDuration)
+	}
+}
+
+func (h *simulatorAPIHandler) terminateContainer(id string, w http.ResponseWriter) {
+	node, ok := h.nodes[id]
+	delete(h.nodes, id) // Almost correct, removal may fail. Lock is too expensive though
+	delete(h.nodesTimeout, id)
+
+	if !ok {
+		h.logger.Error("unknown client deletion requested", "id", id)
+		http.Error(w, "not found", http.StatusNotFound)
```

Comment:
Forgot a `nil`-check for `w` here?

Reply:
I added the nil check. For some reason I thought if it was a nil http.readwriter it would just not write the error and pass on it's way. To be honest I'm not entirely sure that it won't, but this is probably safer.